src/atsamd/timer.c

```diff
@@ -11,19 +11,40 @@
 #include "internal.h" // enable_pclock
 #include "sched.h" // DECL_INIT
 
+#if CONFIG_MACH_SAMC21
+ #define TCp TC0
+ #define TCp_IRQn TC0_IRQn
+ #define TCp_GCLK_ID TC0_GCLK_ID
+ #define ID_TCp ID_TC0
+ #define TCd_GCLK_ID TC1_GCLK_ID
+ #define ID_TCd ID_TC1
+#else
+ #define TCp TC4
+ #define TCp_IRQn TC4_IRQn
+ #define TCp_GCLK_ID TC4_GCLK_ID
+ #define ID_TCp ID_TC4
+ #define TCd_GCLK_ID TC3_GCLK_ID
+ #define ID_TCd ID_TC3
+#endif
+
 // Set the next irq time
 static void
 timer_set(uint32_t value)
 {
-    TC4->COUNT32.CC[0].reg = value;
-    TC4->COUNT32.INTFLAG.reg = TC_INTFLAG_MC0;
+    TCp->COUNT32.CC[0].reg = value;
+    TCp->COUNT32.INTFLAG.reg = TC_INTFLAG_MC0;
 }
 
 // Return the current time (in absolute clock ticks).
 uint32_t
 timer_read_time(void)
 {
-    return TC4->COUNT32.COUNT.reg;
+#if CONFIG_MACH_SAMC21
+    TCp->COUNT32.CTRLBSET.reg |= TC_CTRLBSET_CMD(TC_CTRLBCLR_CMD_READSYNC_Val);
+    while (TCp->COUNT32.SYNCBUSY.reg & TC_SYNCBUSY_COUNT)
+        ;
+#endif
```

Comment:
Can you explain what this does?  The `timer_read_time()` is a "fast path" function - overhead here will have notable impact on performance and benchmark results.  Is there a way to just read the counter value without having to perform a write (or at least without having to perform a read-modify-write and a loop)?

Reply:
There is a note hidden in the data sheet under the 32 bit register summary for this register (section 35.7.3.13) that states this synchronization must happen prior to any read access.